primitives/src/v3/mod.rs

```diff
@@ -0,0 +1,1801 @@
+		input: &mut I,
+	) -> Result<Self, parity_scale_codec::Error> {
+		let maybe_magic = <[u8; 4]>::decode(input)?;
+		if maybe_magic != BACKING_STATEMENT_MAGIC {
+			return Err(parity_scale_codec::Error::from("invalid magic string"))
+		}
+
+		Ok(match CompactStatementInner::decode(input)? {
+			CompactStatementInner::Seconded(h) => CompactStatement::Seconded(h),
+			CompactStatementInner::Valid(h) => CompactStatement::Valid(h),
+		})
+	}
+}
+
+impl CompactStatement {
+	/// Get the underlying candidate hash this references.
+	pub fn candidate_hash(&self) -> &CandidateHash {
+		match *self {
+			CompactStatement::Seconded(ref h) | CompactStatement::Valid(ref h) => h,
+		}
+	}
+}
+
+/// `IndexedVec` struct indexed by type specific indices.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq))]
+pub struct IndexedVec<K, V>(Vec<V>, PhantomData<fn(K) -> K>);
+
+impl<K, V> Default for IndexedVec<K, V> {
+	fn default() -> Self {
+		Self(vec![], PhantomData)
+	}
+}
+
+impl<K, V> From<Vec<V>> for IndexedVec<K, V> {
+	fn from(validators: Vec<V>) -> Self {
+		Self(validators, PhantomData)
+	}
+}
+
+impl<K, V> FromIterator<V> for IndexedVec<K, V> {
+	fn from_iter<T: IntoIterator<Item = V>>(iter: T) -> Self {
+		Self(Vec::from_iter(iter), PhantomData)
+	}
+}
+
+impl<K, V> IndexedVec<K, V>
+where
+	V: Clone,
+{
+	/// Returns a reference to an element indexed using `K`.
+	pub fn get(&self, index: K) -> Option<&V>
+	where
+		K: TypeIndex,
+	{
+		self.0.get(index.type_index())
+	}
+
+	/// Returns number of elements in vector.
+	pub fn len(&self) -> usize {
+		self.0.len()
+	}
+
+	/// Returns contained vector.
+	pub fn to_vec(&self) -> Vec<V> {
+		self.0.clone()
+	}
+
+	/// Returns an iterator over the underlying vector.
+	pub fn iter(&self) -> Iter<'_, V> {
+		self.0.iter()
+	}
+
+	/// Returns a mutable iterator over the underlying vector.
+	pub fn iter_mut(&mut self) -> IterMut<'_, V> {
+		self.0.iter_mut()
+	}
+
+	/// Creates a consuming iterator.
+	pub fn into_iter(self) -> IntoIter<V> {
+		self.0.into_iter()
+	}
+
+	/// Returns true if the underlying container is empty.
+	pub fn is_empty(&self) -> bool {
+		self.0.is_empty()
+	}
+}
+
+/// The maximum number of validators `f` which may safely be faulty.
+///
+/// The total number of validators is `n = 3f + e` where `e in { 1, 2, 3 }`.
+pub fn byzantine_threshold(n: usize) -> usize {
+	n.saturating_sub(1) / 3
+}
+
+/// The supermajority threshold of validators which represents a subset
+/// guaranteed to have at least f+1 honest validators.
+pub fn supermajority_threshold(n: usize) -> usize {
+	n - byzantine_threshold(n)
+}
+
+/// Information about validator sets of a session.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq))]
+pub struct SessionInfo {
+	/****** New in v3 *******/
+	/// Executor parameter set for the session.
+	pub executor_params: ExecutorParams,
```

Comment:
Dumb question: does making a breaking change to `ExecutorParams` (e.g. removing a variant from `ExecutionParam`) make a breaking change to `SessionInfo`? It would be cool if `ExecutorParams` were self-versioned and backwards compatible. I think that was the goal with untyped-map design?

Reply:
We were sorting it out for some time (see discussions [here](https://github.com/paritytech/polkadot/pull/6312#pullrequestreview-1190090425) and [here](https://github.com/paritytech/polkadot/pull/6161/files/cdd62e7156a596349e5a7b1a50cf44bfaa6a2b9c#r1053590905)). Yes, we shouldn't ever remove variants from `ExecutorParams`, otherwise encoding would become broken. That's a bad idea anyway, as old executor params may be required at some point. Adding new variants shouldn't break anything except for nodes that did not upgrade in time, and they'll get a decoding failure. But they wouldn't be able to handle new parameters even if they would have been encoded into an untyped map, and using only part of them (decoding only half of the environment, as @eskimor put it) doesn't seem to be valid.
So it looks like it's solid now, or I want to believe so ;)

Reply:
Sure, if we introduce a param to `ExecutionParams`, the old nodes should not be able to validate a candidate, so no backing, approval-voting or disputes. However, they still should be able to participate in availability, produce blocks and vote for finality. So my point is `SessionInfo` should still be decodable (all but `ExecutorParams`).

Reply:
Also av-recovery could be used by collators which might be slower to upgrade than validators.

Reply:
Yeah, now I get it, it's a really good point!
@eskimor, that's what we both may have missed in the original idea of weakly-typed implementation. What do you think?

Reply:
This is a very good point, but I am a bit reluctant to delay this PR even further. Having validators only doing a fraction of their duties, is hardly worth the complication, I think. Collators are a bigger concern, buut: Shouldn't collators also care about the execution environment the validators are using? I mean they ought to provide something that is valid, so they should know ..

Reply:
What is a bit of a nuisance though, is that we have to bump the `SessionInfo` version every time, we add a new variant to execution environment parameters. :disappointed: 

This can be fixed in two ways:

1. Don't make EE part of `SessionInfo`, but have it separate.
2. Make it an opaque type a the level of `SessionInfo` and only fully decode it where it is actually needed.

1 ... would be the "proper" fix, but likely requires more changes and also comes at the downside that we lose the infrastructure like a persistent `RollingSessionWindow`.
2 ... would likely require less changes and we still have availability guaranteed by `RollingSessionWindow`.

But how frequently do we expect to be adding new variants?

Reply:
We've discussed this with @eskimor on a call.
1) We do not expect the set of variants of `ExecutionParams` to change often. Adding a completely new execution environment is a big deal anyway and deserves a migration, and if we add another version of Wasmtime as a new environment, it will share the same semantics.
2) To mitigate corner cases we might not be aware of right now, I've added a placeholder variant, `RawData(Vec<u8>)`, that may be used for a migrationless alteration of the semantics.